File: finova-python-node/backend/app/providers/sectors.py
import asyncio
import httpx
from typing import Dict, List, Any
from app.config import FINNHUB_API_KEYS, logger

SECTOR_ETFS: Dict[str, str] = {
    'XLK': 'Technology',
    'XLV': 'Health Care',
    'XLF': 'Financials',
    'XLY': 'Consumer Discretionary',
    'XLC': 'Communication Services',
    'XLI': 'Industrials',
    'XLP': 'Consumer Staples',
    'XLE': 'Energy',
    'XLU': 'Utilities',
    'XLRE': 'Real Estate',
    'XLB': 'Materials',
}

ETF_SYMBOLS = list(SECTOR_ETFS.keys())

async def fetch_sector_performance(client: httpx.AsyncClient) -> List[Dict[str, Any]]:
    if not FINNHUB_API_KEYS:
        return []
        
    for key_index, key in enumerate(FINNHUB_API_KEYS):
        try:
            async def fetch_one(sym: str) -> Dict[str, Any]:
                url = f"https://finnhub.io/api/v1/quote?symbol={sym}&token={key}"
                res = await client.get(url, timeout=5.0)
                if res.status_code == 429:
                    raise Exception("Rate limited")
                if res.status_code != 200:
                    raise Exception(f"HTTP {res.status_code}")
                data = res.json()
                return {
                    "name": SECTOR_ETFS[sym],
                    "chg": float(data.get("dp") or 0.0)
                }

            tasks = [fetch_one(sym) for sym in ETF_SYMBOLS]
            results = await asyncio.gather(*tasks, return_exceptions=True)
            
            sectors = []
            for r in results:
                if isinstance(r, dict) and r:
                    sectors.append(r)
                    
            if sectors:
                return sectors
        except Exception as e:
            logger.warning(f"Finnhub sectors fetch failed with key index {key_index}: {e}")
            continue
            
    return []

File: lib/sectors.ts
const SECTOR_ETFS: Record<string, string> = {
  'XLK': 'Technology',
  'XLV': 'Health Care',
  'XLF': 'Financials',
  'XLY': 'Consumer Discretionary',
  'XLC': 'Communication Services',
  'XLI': 'Industrials',
  'XLP': 'Consumer Staples',
  'XLE': 'Energy',
  'XLU': 'Utilities',
  'XLRE': 'Real Estate',
  'XLB': 'Materials',
};

const ETF_SYMBOLS = Object.keys(SECTOR_ETFS);

// Static fallback — approximate recent values, shown when all APIs fail
const STATIC_SECTOR_FALLBACK: { name: string; chg: number }[] = [
  { name: 'Technology',             chg:  1.24 },
  { name: 'Health Care',            chg: -0.38 },
  { name: 'Financials',             chg:  0.67 },
  { name: 'Consumer Discretionary', chg:  0.92 },
  { name: 'Communication Services', chg:  1.05 },
  { name: 'Industrials',            chg:  0.31 },
  { name: 'Consumer Staples',       chg: -0.14 },
  { name: 'Energy',                 chg: -1.22 },
  { name: 'Utilities',              chg: -0.55 },
  { name: 'Real Estate',            chg:  0.18 },
  { name: 'Materials',              chg: -0.29 },
];

// In-memory cache: 10 min TTL
let _sectorCache: { data: { name: string; chg: number }[]; expiry: number } | null = null;

export async function fetchSectorPerformance(): Promise<{ name: string; chg: number }[]> {
  // Return from cache if still valid
  if (_sectorCache && _sectorCache.expiry > Date.now()) {
    return _sectorCache.data;
  }

  const keys = [
    process.env.FINNHUB_API_KEY,
    process.env.FINNHUB_API_KEY_2,
  ].filter(Boolean) as string[];

  if (keys.length > 0) {
    for (const key of keys) {
      try {
        const results = await Promise.allSettled(
          ETF_SYMBOLS.map(sym =>
            fetch(`https://finnhub.io/api/v1/quote?symbol=${sym}&token=${key}`, { next: { revalidate: 60 } } as any)
              .then(r => { if (!r.ok) throw new Error(`HTTP ${r.status}`); return r.json(); })
              .then(d => ({ name: SECTOR_ETFS[sym], chg: d.dp ?? 0 }))
          )
        );
        const sectors = results
          .filter((r): r is PromiseFulfilledResult<{ name: string; chg: number }> =>
            r.status === 'fulfilled' && r.value != null
          )
          .map(r => r.value);

        if (sectors.length > 0) {
          _sectorCache = { data: sectors, expiry: Date.now() + 10 * 60 * 1000 };
          return sectors;
        }
      } catch {
        continue;
      }
    }
  }

  // Return cached stale data if available, otherwise static fallback
  if (_sectorCache) return _sectorCache.data;
  return STATIC_SECTOR_FALLBACK;
}
